tools/tag.sh

```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
+# Shell script equivalent of Metamath TAG tool
+# Mario Carneiro 5-Jan-2022
+
+usage() {
+  cat << HELP
+TAG - Like ADD, but restricted to a range of lines
+
+TAG is the same as ADD but has 4 additional arguments that let you
+specify a range of lines.  Syntax:
+  tag.sh BEGSTR ENDSTR STARTMATCH S# ENDMATCH E# < INFILE > OUTFILE
+where
+  INFILE = input file
+  OUTFILE = output file
+  BEGSTR = string to add to beginning of each line
+  ENDSTR = string to add to end of each line
+  STARTMATCH = a string to match; if empty, match any line
+  S# = the 1st, 2nd, etc. occurrence of STARTMATCH to start the range
+  ENDMATCH = a string to match; if empty, match any line
+  E# = the 1st, 2nd, etc. occurrence of ENDMATCH from the
+      start of range line (inclusive) after which to end the range
+Example:  To add "!" to the end of lines 51 through 60 inclusive:
+  tag.sh '' '!' '' 51 '' 10 < a.txt
+Example:  To add "@@@" to the beginning of each line in theorem
+"abc" through the end of its proof:
+  tag.sh '@@@' '' 'abc $p' 1 '$.' 1 < set.mm
+so that later, SUBSTITUTE can be used to affect only those lines.  You
+can remove the "@@@" tags with SUBSTITUTE when done.
+HELP
+}
+
+if [ $# -eq 1 ] && [ "$1" = "-h" -o "$1" = "--help" ]; then usage; exit; fi
+isNum() { case "$1" in '' | *[!0-9]*) return 1;; esac; }
+if [ $# -ne 6 ] || ! isNum "$4" || ! isNum "$6"; then usage; exit 1; fi
+
+awk -v before="$1" -v after="$2" -v start="$3" -v end="$5" \
+  -v startNum=$4 -v endNum=$6 '{
+  if (startNum > 0) {
```

Comment:
startnum == 0 is not a valid input according to the description.  Here it is used as if it starts the range.

Reply:
`startNum` changes during execution, it counts down to zero. I don't think an initial setting of `startNum == 0` is exceptional in any way code-wise, but it would be possible (with some additional code) to specifically exclude it if required.

Reply:
I find it confusing to deal with undocumented input.  Instead of proper error handling I will see some obscure and unexpected  behaviour.  Once I detect that start == 0 means 'from the beginning' I am lured into thinking end == 0 means 'to the end'.  And so on.  Let the script do what the help says it does.  The rest is just confusing.

Reply:
What I mean to say is that I did not think about the start == 0 case at all when writing this (and this is also just a copy of the metamath code logic), so it does whatever the logical extension of the > 0 behavior is.

Reply:
ack